installation_policies/xSDK_spack_variant_guidelines.md

```diff
@@ -0,0 +1,88 @@
+# xSDK Spack Variant Guidelines
+
+_**Document initiated on 08/25/2020 based on [installation policy discussions](https://github.com/xsdk-project/xsdk-community-policies/pull/20)**_
+
+## Motivation
+
+Common Spack [variants](https://spack.readthedocs.io/en/latest/packaging_guide.html#variants) help package authors to build a consistent and compatible software stack, and offers the potential to support automatic variant propagation. These future features may include forwarding variants supplied to the xSDK metapackage, and possibly specification of globally-scoped or xSDK-scoped variants supplied on the installation line.
+
+For example, variant-forwarding would mean that if a user wished to build the xsdk in _**debug**_ mode, a variant passed to the `xsdk` metapackage could be forwarded to all xSDK packages using `spack install xsdk build_type=debug`
+
+## Package author requirements
+
+This document outlines a list of features that must be exposed as Spack variants, when alternative options for these features are available. If the package does not have options for these features, then a Spack variant is not required.
+
+The exact variant names and types (e.g. boolean or multi-value) used are currently given as recommendations. If the recommended variants are inadequate to express a packages options, then this feedback should be provided to the xSDK team so that these guidelines can be improved.
+
+This document may be updated over time, and so should be checked before each xSDK release.
+
+The proposed recommendations for Spack variants given below are based on a review first carried out on August 18th 2020 The reviewing of Spack variants is open-ended. It is documented in the following location: https://github.com/xsdk-project/xsdk-community-policies/wiki/Spack-variant-review.
+
+## List of recommended variants
+
+The following options must be expressed through Spack variants, where relevant.
+
+
+**1. Index size**
+
+This option concerns index sizes (or compile-time determined integer sizes).
+
+We recommend using the boolean `int64` variant name to set index sizes to 64-bits. It is expected that `~int64` will result in 32-bit integer size.
+
+`int64`: Index sizes (or compile-time determined integer sizes) are set to 64 bits.
+
+When different options are required, an alternative variant (either boolean or multi-valued) may be used. Please refer to the [Spack variant review](https://github.com/xsdk-project/xsdk-community-policies/wiki/Spack-variant-review) for guidance.
+
+---
+
+**2. Precision**
+
+This concerns the size of compile-time configured floating-point variables.
+
+We recommend using multi-valued variant `precision`.
+
+`precision`: The multi-valued variant `precision` can accept multiple precision descriptions. The names `single` and `double` are recommended for 32-bit/64-bit respectively. Other options may include `half` for 16-bit, `extended` for 80-bit or `quad` for 128-bit.
```

Comment:
Are there any packages other than PETSc that also make a distinction between real and complex? Would they encode this via `precision` or some other mechanism?

Reply:
Looks like petsc and trilinos have a complex option, defined as a separate variant. Triilinos considers both when setting an option:

    define('Tpetra_INST_COMPLEX_DOUBLE', complex_s),
    define('Tpetra_INST_COMPLEX_FLOAT', float_s and complex_s),

My instinct is that that its a separate variant. (a complex also has a precision).

Its possible to have multi-value variants that allow more than one option, but I generally think this is best when all permutations make sense (like which cuda architectures to build for). It keeps the information together.

Let me know if you disagree @keitat @balay 

We should maybe think about mixed precision - that thats another thread.

Reply:
We should probably add the option `complex` to the variant guidelines, it is meant as an extendable list. But may be in a follow up PR.

Reply:
For real customers of Trilinos, it is critical to have fine-grained control of what scalar instantiations are done.  For example, just because the customer requires scalar types `float` and `double` does not mean they want `complex<float>` and `complex<double>`.  The extra build time and the extra binary sizes can be too much for them (when they are already on the edge of size limits on a given machine).

Reply:
So in terms of Spack variants, maybe a multi-value variant for complex?

Reply:
heFFTe, MAGMA, PLASMA, and SLATE build in multiple precisions inside a single library: at minimum both real and complex domains, IEEE 754 32-bit and 64-bit floating-pint precision. Some of these libraries have support for IEEE 754 16-bit format. And there are some that build for Google's BrainFloat 16 which is 7+8 bits (same exponent as 32-bit format). Also note that this will likely get more complicated as NVIDIA now supports 7 different real floating-point precisions in hardware. AMD and Intel follow the suit by supporting some of them in their accelerators and ARM has some of them in their CPU processors. Focusing on IEEE-only formats might be limiting for the Exascale hardware and the libraries already in xSDK.

Reply:
I will make this an item for the meeting. I am intereseted in what cmake options are used in these packages, and whether there is commonality.

Other variants could be added in addition to `precision`, eg. `complex_precision`. Other options can also be added to the multi-value variant, like `mixed` or `mixed_64_32`. This might have implications for variant forwarding (maybe we will want a ranked choice variant forwarding?).

For now we need to decide whether `precision` is useful as a variant on its own, or whether it needs a qualifier, or is better removed.

Reply:
@shuds13 There is no option for precision. What I was trying to say is that the 4 libraries, that I mentioned above, build two precisions in two domains and there is no way to enable/disable/customize this behavior. So handling of precisions and domains is implicit. It might be a side-note for this discussion and the note is: some packages handle precision implicitly and will likely do so in the future. This is quite different from packages that build in one precision only that can be selected at configuration time. Most FFT libraries do so, for example FFTW can be built for two different precision and then dependent packages have to deal with that precisions: see HPCC package in Spack.

Reply:
To complete the record, deal.II builds most things in `float` and `double`, and quite a lot (should be all -- but we sometimes forget to add an explicit template instantiation) in `std::complex<float>` and `std::complex<double>`. As pointed out, the downside is that the shared library is positively huge. The upside is that there are no "variant" libraries that are mutually incompatible.